Architecture.md

```diff
@@ -0,0 +1,81 @@
+
+### Proxies and Garbage Collector (GC) Interoperability
+
+JavaScript has a concept of a 'Proxy' object, which we use in emacs-ng, however this section is about our JS <--> elisp marshalling. We refer to that as 'proxying' within internal documentation. For example, we will discuss how this code actually works:
+
+```js
+const buffer = lisp.get_buffer_create('*foo*');
+lisp.set_current_buffer(buffer);
+lisp.insert("FOOBAR");
+```
+
+#### The Lisp Object
+
+The `lisp` object is a [JavaScript Capital-P Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy). When you access `lisp.get_buffer_create`, it returns a function that does something like:
+
+```js
+(...args) => lisp_invoke('get-buffer-create' ...args);
+```
+
+`lisp_invoke` is a native function defined in javascript.rs. It's a wrapper around ffuncall with logic for object translation. It will also catch any lisp errors and translates them to JavaScript errors. `lisp_invoke` returns a JavaScript object - but this is a 'special object' in that it may have an internal field. Internal fields are data objects that are contained within JavaScript objects that are not accessible by users. This is because LispObjects are just pointers - if we let the user alter pointers, they could cause SEGFAULTS or read arbitrary memory.
+
+If we can parse the result of lisp_invoke as JSON, we do not proxy it. Instead we return the JavaScript equivalent (i.e. a JS string or number). However, if we cannot convert it to JSON, like in the case of a buffer, we return a proxy. Functions (i.e. `(lambda () (...))`) are another special case where additional logic is employed.
+
+lisp_invoke also works in reverse when being called, it will `unproxy` the arguments it is passed in order to further pass them to ffuncall.
+
+We expose a special function called `is_proxy` in order to tell if an object is a proxy.
+
+#### GC Usage
+
+When we create a proxy, we need to properly manage it in the lisp garbage collector. We do not want lisp to GC an object out from underneath us. In order to do this, we need to make two considerations for each direction of JS -> Lisp and Lisp -> JS
+
+To prevent the lisp GC from removing objects that JS has a valid reference to, we include them in a special cons called the `js-retain-map`. The user does not have direct access to this object. Allowing them to access this cons would allow them to mutate it in a way that could lead to use after free bugs.
```

Comment:
"Allowing them to access this cons would allow them to mutate it"

maybe "Access to this cons would allow mutating" ?

Reply:
or maybe combine it with the preceding sentence, this why we wouldn't access and allow multiple times.